Open localhost:5050 automatically when running mirrorful cli

karanvirsb opened the pull request:
Made a simple open browser feature following the create-react-app library. #127

alexdanilowicz commented:
Hey @karanvirsb! Thanks for jumping in, but this isn't what I intended from the issue. I was thinking of adding a line in the mirrorful cli script (code found here: https://github.com/Mirrorful/mirrorful/tree/main/packages/mirrorful/cli) or in the package json, as you have it, that prompts the opening of localhost:5050 even if it's not already running.

in other words:

```
npx mirrorful # should open up localhost:5050 in the browser
```

I believe you code right now will only work if localhost:5050 is already running?

karanvirsb commented:
Hey, @alexdanilowicz I followed the create-react-app react-scripts here: https://github.com/facebook/create-react-app/blob/d960b9e38c062584ff6cfb1a70e1512509a966e7/packages/react-dev-utils/openBrowser.js#L140. I found that they used the open library to help them open the browser. If I move the node ./openBrowser.mjs from the script and add it to a different line in the package.json like this:  

```javascript
"scripts": {
    ...
    "openBrowser": "node ./openBrowser.mjs",
    ...
}
```

It will still open the browser and show this: 

![image](https://user-images.githubusercontent.com/71354242/224543805-fc734528-17c6-4198-846a-d2e606252316.png)

karanvirsb commented:
Hey, @alexdanilowicz I have moved openBrowser.ts to the helpers in the cli and added it to the Mirrorful cli init script and it seems to work well. If we moved around the openBrowser function within the init script it determines how early the browser will open.

alexdanilowicz commented on packages/mirrorful/cli/init.ts:
```diff
@@ -31,7 +32,7 @@ export async function init({
 
   const port = 5050 // don't hard code this
-
+  await openBrowser({ url: 'http://localhost', port })
```
Awesome! I think this line should be moved to the end. I think it should be on line 107 instead, after ensuring there are no errors

karanvirsb replied:
If we move it all the way to the bottom it seems to not open the browser.

alexdanilowicz replied:
Oh interesting, makes sense

alexdanilowicz commented on packages/mirrorful/cli/helpers/openBrowser.ts:
```diff
@@ -0,0 +1,12 @@
+    console.log(`Opening up browser to ${url}:${port}`)
+    const res = await open(`${url}:${port}`)
+
+    console.log('🚀 ~ file: openBrowser.ts:8 ~ openBrowser ~ res:', res)
```
```suggestion
    console.log('🚀 Opening Mirrorful in your browser! ', res)
```

karanvirsb replied:
I think maybe we should just remove the result variable from the log as it might bloat up the console.

alexdanilowicz commented on packages/mirrorful/cli/helpers/openBrowser.ts:
```diff
@@ -0,0 +1,12 @@
+
+    console.log('🚀 ~ file: openBrowser.ts:8 ~ openBrowser ~ res:', res)
+  } catch (error) {
+    console.log('🚀 ~ file: openBrowser.ts:10 ~ openBrowser ~ error:', error)
```
```suggestion
    console.log('Go to localhost:5050 to open Mirrorful. Could not open automatically: ', error)
```

alexdanilowicz reviewed (commented):
few nit comments

alexdanilowicz commented:
Hey @karanvirsb awesome! incredible, thanks so much for your time and help with this! Left a few comments.

alexdanilowicz commented:
By the way, @karanvirsb, are you in our [Slack community](https://join.slack.com/t/mirrorful/shared_invite/zt-1ps2xtxh0-2NaixFfFzSKZbr5gw_AHfA)? 

Gave you a shoutout here:
![image](https://user-images.githubusercontent.com/29822597/224564686-cab390aa-dcb1-4dcf-85ca-b5f0e6d2da9c.png)

alexdanilowicz commented:
Hmmm, I think the lag of the site not being available thought is not a good user experience.

Demostration here:
https://www.loom.com/share/9734c09251f347deba19540086ca3e33

Really appreciate the hard work on this, but I think we either 

1) Need to ensure it doesn't even pop up when not available.. perhaps put it in some sort of timeout? That might be too hacky though. But I think it would be good to see the console output too.
2) If we can that to work, not a big deal as we are probably going to move to a cloud editor soon.

karanvirsb commented:
I think the delay is mostly due to the next dev and the spawn child sync. I tried fixing it by making the spawn async. setTimeout might be too hacky since you need to know how long the next dev takes to run.

alexdanilowicz commented:
Nice, I like it and think it's a fantastic improvement. I made a few changes to the language that is outputted. Let me know what you think. 

@Teddarific — do you have thoughts here? I would prefer to release `4.0.1` first, and then we merge this change to be released with `4.0.2`.

karanvirsb commented:
> Nice, I like it and think it's a fantastic improvement. I made a few changes to the language that is outputted. Let me know what you think.
> 
> @Teddarific — do you have thoughts here? I would prefer to release `4.0.1` first, and then we merge this change to be released with `4.0.2`.

Nice, I think it looks good, I like the changes!
